importer/Bosnia.py

```diff
@@ -0,0 +1,321 @@
+importer that parses the csv data
+(in several files in a subdirectory called
+upon with --datadir).
+"""
+import argparse
+import os
+import dateparser
+
+import wikidataStuff.wdqsLookup as lookup
+
+from Monument import Monument
+from Uploader import Uploader
+import importer_utils as utils
+
+MAPPINGS_DIR = "mappings"
+REPORTING_DIR = "reports"
+PREVIEW_DIR = "previews"
+
+
+class BosniaMonument(Monument):
+
+    def __init__(self, db_row_dict, mapping, data_files, existing, repository):
+        """
+        Initialize the data object.
+
+        Like Monument.py but with some simplifications
+        re: attribute setting
+        """
+        self.raw_data = db_row_dict
+        self.props = data_files["_static"]["props"]
+        self.adm0 = data_files["_static"]["adm0"]
+        self.sources = data_files["_static"]["sources"]
+        self.common_items = data_files["_static"]["common_items"]
+        for k, v in db_row_dict.items():
+            setattr(self, k.lower().strip().replace(" ", "_"), v)
+        self.data_files = data_files
+        self.construct_wd_item(mapping)
+        self.existing = existing
+        self.repo = repository
+        self.problem_report = {}
+
+    def create_source(self):
+        source_item = "Q42750314"
+        last_edit = "2017-09-08"
+        timestamp = utils.date_to_dict(last_edit, "%Y-%m-%d")
+        prop_stated = self.props["stated_in"]
+        prop_date = self.props["publication_date"]
+        self.source = {"source": {"prop": prop_stated, "value": source_item},
+                "published": {"prop": prop_date,
+                "value": timestamp}, }
+
+    def set_country(self):
+        self.add_statement("country", "Q225", refs=self.source)
+
+    def set_heritage(self):
+        """
+        Set heritage.
+
+        National Monument of Bosnia and Herzegovina (Q12637954)
+        with facultative start date.
+        """
+        heritage = self.mapping["heritage"]["item"]
+        date_parsed = dateparser.parse(self.designated)
+        quals = {}
+        if date_parsed:
+            date_dict = utils.datetime_object_to_dict(date_parsed)
+            quals["start_time"] = {"time_value": date_dict}
+
+        self.add_statement("heritage_status", heritage,
+                           quals=quals, refs=self.source)
+
+    def set_is(self):
+        """
+        Set P31.
+
+        Try to get a specific one via mapping file.
+        Otherwise default of cultural property.
+        """
+        default = self.mapping["default_is"]["item"]
+        type_dict = self.data_files["_static"]["bosnia_types"]
+        type_raw = self.object_type.strip()
+        type_match = utils.get_item_from_dict_by_key(
+            dict_name=type_dict,
+            search_in="itemLabel",
+            search_term=type_raw)
+        if len(type_match) == 1:
+            to_add = type_match[0]
+        else:
+            to_add = default
+
+        self.add_statement("is", to_add, refs=self.source)
```

Comment:
add `, if_empty=True)` to not add when there are pre-existing P31s

Reply:
:white_check_mark: